omniledger/contracts/coins.go

```diff
@@ -123,35 +124,37 @@ func ContractCoin(cdb ol.CollectionView, inst ol.Instruction, c []ol.Coin) (sc [
 				return
 			}
 
-			targetCoin := newSafeUint64(v)
-			targetCoin, err = targetCoin.add(coinsArg)
+			var targetCI CoinInstance
+			err = protobuf.Decode(v, &targetCI)
+			if err != nil {
+				return nil, nil, errors.New("couldn't unmarshal target account: " + err.Error())
+			}
+			targetCI.Balance, err = safeAdd(targetCI.Balance, coinsArg)
 			if err != nil {
 				return
 			}
-			var w bytes.Buffer
-			binary.Write(&w, binary.LittleEndian, targetCoin)
+			targetBuf, err := protobuf.Encode(&targetCI)
+			if err != nil {
+				return nil, nil, errors.New("couldn't marshal target account: " + err.Error())
+			}
 
 			log.Lvlf3("transferring %d to %x", coinsArg, target)
 			sc = append(sc, ol.NewStateChange(ol.Update, ol.NewInstanceID(target),
-				ContractCoinID, w.Bytes(), did))
+				ContractCoinID, targetBuf, did))
 		case "fetch":
 			// fetch removes coins from the account and passes it on to the next
 			// instruction.
-			if coinsArg > uint64(coinsCurrent) {
-				err = errors.New("not enough coins in instance")
-				return
-			}
-			coinsCurrent, err = coinsCurrent.sub(coinsArg)
+			ci.Balance, err = safeSub(ci.Balance, coinsArg)
 			if err != nil {
 				return
 			}
-			cOut = append(cOut, ol.Coin{Name: CoinName, Value: coinsArg})
+			cOut = append(cOut, ol.Coin{Name: ol.NewInstanceID(ci.Type), Value: coinsArg})
```

Comment:
NewInstanceID converts the first 32 bytes of in into an InstanceID. What if ci.Type is shorter than 32 bytes? I actually had a todo written down a long time ago to understand what this CoinName thing is.

Reply:
Don't know what I've been thinking. It should simply store a `coin`, not invent a new structure and then convert between the two...